### DISCUSSION TITLE
🆕 New Source: SAP HANA

### DISCUSSION THREAD

**Discussion by jgngo** [airbytehq/airbyte#31796] [New Connector Request]
## Tell us about the new integration you’d like to have
* SAP HANA to Postgres

## Describe the context around this new integration
* Creating a finance data mart
* Sync every hour

## Describe the alternative you are considering or using
* Prefect or Airflow

https://help.sap.com/viewer/4fe29514fd584807ac9f2a04f6754767/2.0.03/en-US/20a1569875191014b507cf392724b7eb.html

1. Investigate sandbox environment 
- https://developers.sap.com/tutorials/hxe-ua-install-using-docker.html
- or create a sandbox in cloud.
2. Generate new connector and implement connections https://developers.sap.com/tutorials/hana-clients-jdbc.html 
3. Create unit test 
4. Integration test 
5. Comprehensive:
- Investigate and prepare use cases:  
https://help.sap.com/viewer/6811c09434084fd1bc4f40e66913ce11/7.4.19/en-US/4ec07a0b6e391014adc9fffe4e204223.html
- Implement comprehensive test. 
6: Documentation, Prepare pull request, pass all checks. 

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368132385412) by [Unito](https://www.unito.io)

**Comment by marcosmarxm** [comment]
Thanks @jgngo I added your request to the roadmap. We just released Airbyte CDK that improves building new connectors. Are you willing to build this connector? (it's not a problem if not)

**Comment by jgngo** [comment]
Would love to build one, if I get time to do it.

**Comment by sherifnada** [comment]
blocked on setting up a SAP Hana instance in AWS/GCP due to the instance being very cumbersome to setup.

**Comment by trymzet** [comment]
Sorry if this is SPAM, but I randomly came across this article which describes setting up Hana in a docker container: https://feitam.es/how-implement-cdc-change-data-capture-of-database-hana-apache-kafka-connect-sap-connector-incremental-mode-to-postgresql/. You may find it useful.

**Comment by sherifnada** [comment]
Thanks @Trymzet ! We'll take a look at it soon once we get to implementing this connector

**Comment by jomach** [comment]
I'm working since some months on building a kafka source connector for SAP. Be aware you will have Problems on calculating deltas. As SAP Tables do not have "updatedAt" column for example.

**Comment by jomach** [comment]
You can go over triggers but you could go over ODP queues.  See https://init-software.de/odpsc/#/

**Comment by sherifnada** [comment]
Thanks for the heads up @jomach

**Comment by ofranz** [comment]
Any updates or timeline on this connector?

**Comment by matheus-rossi** [comment]
Hello! I've worked with SAP HANA recently, developing an incremental data pipeline to extract data from HANA.

As @jomach commented, the absence of columns like "updatedAt" was a challenge in the beginning, but I found an internal column present in all SAP HANA tables, which is ("$rowid$").

Based on this community post: https://blogs.sap.com/2023/03/12/rowid-function-in-HANA/

I documented this in my docs; it may be useful: https://matheusrossi.com/development/erp/sap-hana/

@sherifnada  Maybe this ("$rowid$") column could be used to perform incremental reads from tables...

**Comment by QuanLab** [comment]
`$rowid$` seems to be working fine in this case. Is there anyone working on this?

**Comment by jrosskopf** [comment]
If you guys are not necessarily interested in SAP HANA via the database, but higher level integration on **SAP ERP tables**, **BAPIs**, but also **SAP Business Warehouse (BW)** cubes and querys and **ODP data sources** (e.g. CDS views, but also HANA tables), then maybe have a look at: https://github.com/datazoode/erpl-airbyte-connectors

We just did an experiment with Airbyte to see whether there is demand from the community. In case you are interested, open an issue in the repo or directly contact us via: https://erpl.io/contact.html

**Comment by jomach** [comment]
Update: ODP data sources were killed by SAP and cannot be used by external extractors anymore.  So I guess we will need to go with https://github.com/datazoode/erpl-airbyte-connectors

**Reply by jrosskopf** [reply]
@jomach, thank you for your reply, and the plug  the [ERPL airbyte connector](https://github.com/DataZooDE/erpl-airbyte-connectors). Just to clarify, ERPL also can replicate data via the ODP RFC interface, but also supports BAPIs, read table or read from BW cubes / queries via BICS. 

Maybe also 2cents on the *killing of ODP*. They literally say, that the *do not support* ODP via RFC anymore and suggest ODP OData for 3rd party integrations. Of course their own tools will continue using ODP RFC, as it is currently the only reliable, scalable data replication method on SAP application server level. However, what is interesting is:  

- That it is not clear how this reduction of scope of the RFC interface interferes with the exiting contracts of customers. And
- What the many other third party integrations like Azure Data Factory, Matillion, Google Cloud Data Fusion, Theobald, Informatica, ... to just name a few ... will do.

From SAPs side is is just a way to push their Datasphere Product, which at the current state is super immature and has ridiculously expensive data egress rates.

**Reply by jomach** [reply]
IMHO is ridiculous that they force clients that put clients data into the SAP BW and then SAP say oh no you cannot pull the data out...

**Comment by jomach** [comment]
Currently I'm using the field REQTSN on SAP BW. This is a incremental Timestamp

**Comment by jomach** [comment]
@jgngo does the erpl support incremental loading ?

**Comment by ngohuuduc** [comment]
Will this connector run on JDBC ? Since we want to replace our current JDBC legacy pipeline

**Reply by jomach** [reply]
What is the issue from jdbc for you?

**Reply by ngohuuduc** [reply]
Ah. 
We are looking to replace our current ETL solution with Airbyte from SAP HANA (JDBC) and hope this will work soon.

**Comment by venkatamandavilli-code** [comment]
Thanks for laying out the context and the steps you’ve already explored — this is a great overview of what’s involved in building a SAP HANA → Postgres integration.

A connector for SAP HANA would definitely be valuable, especially for teams building finance or operational data marts where hourly syncs are common. Since SAP HANA isn’t currently supported out of the box, your plan to investigate the sandbox environment, validate the JDBC connection, and build out unit and integration tests is exactly the right approach.

If you decide to move forward with generating a new connector, the links you shared (HANA client setup, JDBC docs, and the SAP data model references) will be very helpful for shaping the schema and test coverage. Once you have a working prototype, feel free to open a PR — even if it’s still early — and we can help review the implementation, testing strategy, and documentation.

Looking forward to seeing what you build.

**Comment by jacopobonomi** [comment]
Hi all! I've built an open-source SAP HANA source for Airbyte (Python CDK + `hdbcli`, MIT), and I'd like to contribute it as a community connector. This also covers #33162.

**Code:** https://github.com/jacopobonomi/source-sap-hana (standalone) · monorepo branch: https://github.com/jacopobonomi/airbyte/tree/jacopobonomi/source-sap-hana-community

### Why

The docs for the SAP HANA Enterprise connector say *"Airbyte no longer sells this connector"*, so Airbyte OSS users now have no way
to read from SAP HANA, the database behind every S/4HANA and BW/4HANA system. I built it for our own S/4HANA → warehouse pipelines
(the same SAP HANA → Postgres use case as this discussion).

### What it does

- Discovery from the `SYS` catalog (tables, optionally views, primary keys) with schema and `LIKE` filters. An S/4HANA schema has 100k+ tables.
- **Resumable full refresh**: keyset pagination on the primary key, with a checkpoint after every page.
- **Cursor-based incremental**: a checkpoint is only emitted once every row sharing a cursor value has been read, so none are lost.
- **Per-stream settings**: SQL row filter, default cursor, and primary key override (e.g. `ACDOCA` has no PK constraint in the catalog).
- Reconnect-and-resume on transient `hdbcli` errors (`-10807 Connection down`, …), Airbyte's standard SSH tunnel, TLS/HANA Cloud/tenant DBs,
  protocol compression, opt-in concurrent stream reads, row estimates, and check-time validation of the row filters.

### How it was validated

- 90 unit tests, including the Airbyte CDK standard source tests, run against an in-process fake HANA (SQLite with `SYS` and schema databases attached).
  The fake injects real `hdbcli` connection errors to test reconnect and resume, so no server is needed in CI.
- Integration tests against a production S/4HANA: every stream's values checked against the discovered schema, and a keyset read of `ACDOCA`
  (1.6 billion rows) reconciled with HANA's `COUNT(*)` and `SUM(ABS(HSL))`.
- Running in Airbyte OSS 1.x (`abctl`, arm64), with syncs completing successfully.

### Questions for the team

1. **Naming:** is `airbyte/source-sap-hana` with display name "SAP HANA (Community)" fine, given the enterprise connector is `source-sap-hana-enterprise`?
2. **CI:** there is no public HANA sandbox. Are unit tests against the fake HANA enough, or should CI spin up SAP HANA Express (`saplabs/hanaexpress`, it needs about 8 GB of RAM)?
3. Anything else you'd need before I open the PR?

I'm happy to maintain it.
